### PR TITLE
Tranfer aliases to a new mailbox when deleting mailboxes

### PR DESCRIPTION
**Pull request by acasajus**
Improvement on https://github.com/simple-login/app/pull/1457 
I had to do it in a different MR because I can't push to @mlec1 's branch

### CODE REVIEW

**Comment by mlec1** [comment]
Hello, 

Sorry for my late reaction and that you had to take over this.

Just a small comment concerning the `npm ci` in the Dockerfile, because there were the question about it on my branch:
- `npm install` may write to **package-lock.json** if a dependency of **package.json** is missing or a newer version is available
- `npm ci` will install all the packages directly from the **package-lock.json** file

Here are some better explanation: https://stackoverflow.com/questions/52499617/what-is-the-difference-between-npm-install-and-npm-ci

I had the issue once in another project where the **npm install** was installing newer dependencies which were breaking the project. After searching about the issue, we discovered the **npm ci** command line, and we went for it.

I thought it might be a safer install method for you too, but I let you the decision to keep it or not, I am far from being a npm expert

**Comment by acasajus** [comment]
I've added your fix to the Dockerfile in a different MR https://github.com/simple-login/app/pull/1543 . Thanks!

**Review by cquintana92 — APPROVED** [review]

File: app/api/views/mailbox.py [hunk 1]
```diff
@@ -78,24 +78,52 @@ def delete_mailbox(mailbox_id):
...
+        (optional) transfer_aliases_to: in body. Id of the new mailbox for the aliases.
+                                        If omitted or the value is set to -1,
+                                        the aliases of the mailbox will be deleted too.
     Output:
         200 if deleted successfully
 
     """
     user = g.user
-    mailbox = Mailbox.get(mailbox_id)
+    mailbox = Mailbox.get(id=mailbox_id)
```
Comment (nguyenkims) [inline app/api/views/mailbox.py:89]: Nits: we can remove the `id=` part here

File: app/dashboard/views/mailbox.py [hunk 2]
```diff
@@ -40,28 +47,53 @@ def mailbox_route():
...
             if mailbox.id == current_user.default_mailbox_id:
                 flash("You cannot delete default mailbox", "error")
                 return redirect(url_for("dashboard.mailbox_route"))
 
+            transfer_mailbox_id = delete_mailbox_form.transfer_mailbox_id.data
+            if transfer_mailbox_id and transfer_mailbox_id > 0:
+                transfer_mailbox = Mailbox.get(transfer_mailbox_id)
+
+                if not transfer_mailbox or transfer_mailbox.user_id != current_user.id:
+                    flash("You must transfer the aliases to a mailbox you own.")
```
Comment (nguyenkims) [inline app/dashboard/views/mailbox.py:72]: `flash` requires a second argument for the message level (warning, info, etc)

File: app/dashboard/views/mailbox.py [hunk 3]
```diff
@@ -112,12 +147,12 @@ def mailbox_route():
                 elif not email_can_be_used_as_mailbox(mailbox_email):
                     flash(f"You cannot use {mailbox_email}.", "error")
                 else:
-                    new_mailbox = Mailbox.create(
+                    transfer_mailbox = Mailbox.create(
```
Comment (nguyenkims) [inline app/dashboard/views/mailbox.py:150]: We shouldn't rename `new_mailbox` to `transfer_mailbox` here right?

File: job_runner.py [hunk 4]
```diff
@@ -124,6 +124,58 @@ def welcome_proton(user):
...
+            alias_transferred_to = transfer_mailbox.email
+
+            for alias in mailbox.aliases:
+                if alias.mailbox_id == mailbox.id:
+                    alias.mailbox_id = transfer_mailbox.id
+                    if transfer_mailbox in alias._mailboxes:
+                        alias._mailboxes.remove(transfer_mailbox)
+                else:
+                    alias._mailboxes.remove(mailbox)
+                    if transfer_mailbox not in alias._mailboxes:
```
Comment (nguyenkims) [inline job_runner.py:147]: can that happen that `alias.mailbox_id = transfer_mailbox_id`?

File: app/api/views/mailbox.py (hunk 1 above)
Comment (mlec1) [reply app/api/views/mailbox.py:89]: Done here https://github.com/simple-login/app/pull/1547

File: app/dashboard/views/mailbox.py (hunk 2 above)
Comment (mlec1) [reply app/dashboard/views/mailbox.py:72]: Done here https://github.com/simple-login/app/pull/1547

File: app/dashboard/views/mailbox.py (hunk 3 above)
Comment (mlec1) [reply app/dashboard/views/mailbox.py:150]: I agree. Done here https://github.com/simple-login/app/pull/1547

File: job_runner.py (hunk 4 above)
Comment (mlec1) [reply job_runner.py:147]: I am not sure to understand your question:

If the mailbox we want to delete, is the "main" mailbox of the alias, we set the "main" mailbox to the **transfer_mailbox** and the value is saved with Session.commit()

Comment (acasajus) [reply job_runner.py:147]: It's validated in the request before creating the job that they aren't the same.

Comment (mlec1) [reply job_runner.py:147]: Ha ok, I understand the question now. Yeah, I agree with @acasajus.

**Pull request merged**